From b3982489d46d3bab5722122d31ae977bf89c543c Mon Sep 17 00:00:00 2001
From: Oliver Beard <olib141@outlook.com>
Date: Wed, 25 Mar 2026 20:40:11 +0000
Subject: [PATCH 1/3] BACKPORT: {annotation/image}document: Use QImageReader to
 respect image transformation

QImageReader can respect image transformations (such as rotation) that are hinted in image metadata, for example, a rotated JPEG. Koko, for instance, already does this, and so should we.

BUG: 516546
(cherry picked from commit d1ae60cde0c821ce207a52069f4b43ad77819991)
---
 src/annotations/annotationdocument.cpp | 7 +++++--
 src/imagedocument.cpp                  | 6 +++++-
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/src/annotations/annotationdocument.cpp b/src/annotations/annotationdocument.cpp
index dcc6f52cb876..a3572c91401b 100644
--- a/src/annotations/annotationdocument.cpp
+++ b/src/annotations/annotationdocument.cpp
@@ -8,6 +8,7 @@
 #include "utils.h"
 
 #include <QGuiApplication>
+#include <QImageReader>
 #include <QPainter>
 #include <QPainterPath>
 #include <QQuickItem>
@@ -179,12 +180,14 @@ void AnnotationDocument::setBaseImage(const QImage &image)
 
 void AnnotationDocument::setBaseImage(const QString &path)
 {
-    setBaseImage(QImage(path));
+    QImageReader reader(path);
+    reader.setAutoTransform(true);
+    setBaseImage(reader.read());
 }
 
 void AnnotationDocument::setBaseImage(const QUrl &localFile)
 {
-    setBaseImage(QImage(localFile.toLocalFile()));
+    setBaseImage(localFile.toLocalFile());
 }
 
 void AnnotationDocument::cropCanvas(const QRectF &cropRect)
diff --git a/src/imagedocument.cpp b/src/imagedocument.cpp
index 43206979aec9..fe884021bcfc 100644
--- a/src/imagedocument.cpp
+++ b/src/imagedocument.cpp
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: LGPL-2.1-or-later
  */
 
+#include <QImageReader>
+
 #include "imagedocument.h"
 
 #include "commands/cropcommand.h"
@@ -15,7 +17,9 @@ ImageDocument::ImageDocument(QObject *parent)
     : QObject(parent)
 {
     connect(this, &ImageDocument::pathChanged, this, [this](const QUrl &url) {
-        m_image = QImage(url.isLocalFile() ? url.toLocalFile() : url.toString());
+        QImageReader reader(url.isLocalFile() ? url.toLocalFile() : url.toString());
+        reader.setAutoTransform(true);
+        m_image = reader.read();
         m_edited = false;
         Q_EMIT editedChanged();
         Q_EMIT imageChanged();

base-commit: dc18bdf112c2959d515608fb524207b7d2728c34
-- 
2.52.0

